File: python/src/susudigital/batch.py
"""
Batch processing utilities for the Susu Digital Python SDK.

:class:`BatchProcessor` allows high-volume operations to be split into
configurable batches and executed with configurable concurrency.

Usage::

    from susudigital.batch import BatchProcessor

    processor = BatchProcessor(client, batch_size=100)

    customers_data = [
        {"first_name": "John", "last_name": "Doe", "phone": "+233XXXXXXXXX"},
        ...
    ]

    results = processor.customers.create_batch(customers_data)

    for result in results:
        if result.success:
            print(f"Created: {result.data.id}")
        else:
            print(f"Failed: {result.error}")
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Any, Callable, Generic, List, Optional, TypeVar

logger = logging.getLogger(__name__)

T = TypeVar("T")


@dataclass
class BatchResult(Generic[T]):
    """Result of a single item in a batch operation."""

    success: bool
    data: Optional[T] = None
    error: Optional[Exception] = None
    index: int = 0


@dataclass
class BatchResults(Generic[T]):
    """Aggregate result of a full batch operation."""

    results: List[BatchResult[T]] = field(default_factory=list)

    @property
    def successful(self) -> List[T]:
        return [r.data for r in self.results if r.success and r.data is not None]

    @property
    def failed(self) -> List[BatchResult[T]]:
        return [r for r in self.results if not r.success]

    @property
    def success_count(self) -> int:
        return sum(1 for r in self.results if r.success)

    @property
    def failure_count(self) -> int:
        return sum(1 for r in self.results if not r.success)

    def __iter__(self):  # type: ignore[override]
        return iter(self.results)


class _BatchServiceWrapper:
    """Internal helper that wraps a service and adds ``create_batch`` / ``list_batch``."""

    def __init__(self, service: Any, batch_size: int) -> None:
        self._service = service
        self._batch_size = batch_size

    def __getattr__(self, name: str) -> Any:
        return getattr(self._service, name)

    def create_batch(self, items: List[Any]) -> BatchResults:
        """Create many resources in batches, collecting successes and failures."""
        results: BatchResults = BatchResults()

        for i in range(0, len(items), self._batch_size):
            chunk = items[i : i + self._batch_size]
            for j, item in enumerate(chunk):
                index = i + j
                try:
                    data = self._service.create(item)
                    results.results.append(BatchResult(success=True, data=data, index=index))
                except Exception as exc:
                    logger.warning("Batch create failed for item %d: %s", index, exc)
                    results.results.append(
                        BatchResult(success=False, error=exc, index=index)
                    )

        return results


class BatchProcessor:
    """Process SDK operations in configurable batches.

    Args:
        client: A :class:`~susudigital.SusuDigitalClient` instance.
        batch_size: Maximum items per batch (default: ``100``).
    """

    def __init__(self, client: Any, batch_size: int = 100) -> None:
        self._client = client
        self._batch_size = batch_size

    @property
    def customers(self) -> _BatchServiceWrapper:
        return _BatchServiceWrapper(self._client.customers, self._batch_size)

    @property
    def transactions(self) -> _BatchServiceWrapper:
        return _BatchServiceWrapper(self._client.transactions, self._batch_size)

    @property
    def loans(self) -> _BatchServiceWrapper:
        return _BatchServiceWrapper(self._client.loans, self._batch_size)

File: php/src/Batch.php
<?php

declare(strict_types=1);

namespace SusuDigital;

use SusuDigital\Exceptions\SusuDigitalException;

/**
 * Batch processing utilities for the Susu Digital PHP SDK.
 *
 * BatchProcessor allows high-volume operations to be split into configurable
 * chunks and executed with error isolation – a failure on one item does not
 * abort the remaining items.
 *
 * Usage:
 *
 *   $processor = new BatchProcessor($client, batchSize: 100);
 *
 *   $results = $processor->customers->createBatch([
 *       ['first_name' => 'John', 'last_name' => 'Doe', 'phone' => '+233XXXXXXXXX'],
 *       // …
 *   ]);
 *
 *   foreach ($results as $result) {
 *       if ($result->success) {
 *           echo "Created: " . $result->data->id . PHP_EOL;
 *       } else {
 *           echo "Failed: "  . $result->error->getMessage() . PHP_EOL;
 *       }
 *   }
 */

// ---------------------------------------------------------------------------
// Value objects
// ---------------------------------------------------------------------------

/**
 * Result of a single item within a batch operation.
 *
 * @template T
 */
final class BatchResult
{
    /**
     * @param T|null $data
     */
    public function __construct(
        public readonly bool $success,
        public readonly mixed $data = null,
        public readonly ?\Throwable $error = null,
        public readonly int $index = 0,
    ) {}
}

/**
 * Aggregate result of a full batch operation.
 *
 * @template T
 */
final class BatchResults implements \Countable, \IteratorAggregate
{
    /** @var BatchResult<T>[] */
    private array $results = [];

    /**
     * @param BatchResult<T> $result
     */
    public function add(BatchResult $result): void
    {
        $this->results[] = $result;
    }

    /** @return BatchResult<T>[] */
    public function all(): array
    {
        return $this->results;
    }

    /** @return T[] */
    public function successful(): array
    {
        return array_values(array_map(
            static fn (BatchResult $r) => $r->data,
            array_filter($this->results, static fn (BatchResult $r) => $r->success && $r->data !== null),
        ));
    }

    /** @return BatchResult<T>[] */
    public function failed(): array
    {
        return array_values(array_filter(
            $this->results,
            static fn (BatchResult $r) => !$r->success,
        ));
    }

    public function successCount(): int
    {
        return count(array_filter($this->results, static fn (BatchResult $r) => $r->success));
    }

    public function failureCount(): int
    {
        return count(array_filter($this->results, static fn (BatchResult $r) => !$r->success));
    }

    public function count(): int
    {
        return count($this->results);
    }

    /** @return \ArrayIterator<int, BatchResult<T>> */
    public function getIterator(): \ArrayIterator
    {
        return new \ArrayIterator($this->results);
    }
}

// ---------------------------------------------------------------------------
// Internal service wrapper
// ---------------------------------------------------------------------------

/**
 * Wraps a service and adds createBatch support.
 *
 * @internal
 */
final class BatchServiceWrapper
{
    public function __construct(
        private readonly object $service,
        private readonly int $batchSize,
    ) {}

    /**
     * Proxy unknown methods to the wrapped service.
     *
     * @param mixed[] $args
     */
    public function __call(string $name, array $args): mixed
    {
        return $this->service->$name(...$args);
    }

    /**
     * Create many resources in batches, collecting successes and failures.
     *
     * @param  array<int, array<string, mixed>> $items
     * @return BatchResults<mixed>
     */
    public function createBatch(array $items): BatchResults
    {
        $results = new BatchResults();

        foreach (array_chunk($items, $this->batchSize, preserve_keys: true) as $chunk) {
            foreach ($chunk as $index => $item) {
                try {
                    $data = $this->service->create($item);
                    $results->add(new BatchResult(success: true, data: $data, index: $index));
                } catch (\Throwable $e) {
                    error_log(sprintf(
                        '[SusuDigital] Batch create failed for item %d: %s',
                        $index,
                        $e->getMessage(),
                    ));
                    $results->add(new BatchResult(success: false, error: $e, index: $index));
                }
            }
        }

        return $results;
    }
}

// ---------------------------------------------------------------------------
// Public entry-point
// ---------------------------------------------------------------------------

/**
 * Process SDK operations in configurable batches.
 */
final class BatchProcessor
{
    public function __construct(
        private readonly SusuDigitalClient $client,
        private readonly int $batchSize = 100,
    ) {}

    public function getCustomers(): BatchServiceWrapper
    {
        return new BatchServiceWrapper($this->client->customers, $this->batchSize);
    }

    public function getTransactions(): BatchServiceWrapper
    {
        return new BatchServiceWrapper($this->client->transactions, $this->batchSize);
    }

    public function getLoans(): BatchServiceWrapper
    {
        return new BatchServiceWrapper($this->client->loans, $this->batchSize);
    }

    /**
     * Magic property access: $processor->customers, $processor->loans, etc.
     */
    public function __get(string $name): BatchServiceWrapper
    {
        return match ($name) {
            'customers'    => $this->getCustomers(),
            'transactions' => $this->getTransactions(),
            'loans'        => $this->getLoans(),
            default        => throw new \InvalidArgumentException(
                "BatchProcessor has no service '{$name}'"
            ),
        };
    }
}
